lib/dicon/secure_shell.ex

```diff
@@ -96,16 +104,39 @@ defmodule Dicon.SecureShell do
   end
 
   def copy(conn, source, target) do
-    %{size: size} = File.stat!(source)
-    stream = File.stream!(source, [], div(size, 100))
-    {:ok, channel} = :ssh_sftp.start_channel(conn, [timeout: @timeout])
-    {:ok, handle} = :ssh_sftp.open(channel, target, [:write, :binary], @timeout)
-    Enum.each(stream, fn chunk ->
-      :ok = :ssh_sftp.write(channel, handle, chunk, @timeout)
-      IO.write "."
-    end)
-    IO.puts "\n"
-    :ok = :ssh_sftp.close(channel, handle, @timeout)
-    :ok = :ssh_sftp.stop_channel(channel)
+    res =
+      with {:ok, %File.Stat{size: size}} <- File.stat(source),
+           stream = File.stream!(source, [], div(size, 100)),
+           {:ok, channel} <- :ssh_sftp.start_channel(conn, [timeout: @timeout]),
+           {:ok, handle} <- :ssh_sftp.open(channel, target, [:write, :binary], @timeout),
+           Enum.each(stream, fn chunk ->
+             # TODO: we need to remove this assertion here as well, once we have a
+             # better "streaming" API.
+             :ok = :ssh_sftp.write(channel, handle, chunk, @timeout)
+             IO.write "."
+           end),
+           :ok <- IO.puts("\n"),
+           :ok <- :ssh_sftp.close(channel, handle, @timeout),
+           :ok <- :ssh_sftp.stop_channel(channel),
+        do: :ok
+
+    format_if_error(res)
+  end
+
+  defp format_if_error({:error, reason} = error) when is_binary(reason) do
+    error
+  end
+
+  defp format_if_error({:error, reason}) do
+    inet_reason = :inet.format_error(reason)
+    if inet_reason == 'unknown POSIX error' do
+      {:error, inspect(reason)}
+    else
+      {:error, List.to_string(inet_reason)}
+    end
+  end
+
+  defp format_if_error(not_an_error) do
```

Comment:
I'm not sure why it's better. It may be more readable as it doesn't have an "informal" feeling (like `not_an_error`), so we can change it if you want. :)